libafl/src/stages/tmin.rs

```diff
@@ -0,0 +1,388 @@
+        loop {
+            if i >= num {
+                break;
+            }
+
+            let mut next_i = i + 1;
+            let mut input = base.clone();
+
+            let before_len = input.len();
+
+            state.set_max_size(before_len);
+
+            start_timer!(state);
+            self.mutator_mut().mutate(state, &mut input, i as i32)?;
+            mark_feature_time!(state, PerfFeature::Mutate);
+
+            let corpus_idx = if input.len() < before_len {
+                // run the input
+                let exit_kind = fuzzer.execute_input(state, executor, manager, &input)?;
+                let observers = executor.observers();
+
+                // let the fuzzer process this execution -- it's possible that we find something
+                // interesting, or even a solution
+                let (_, corpus_idx) = fuzzer.process_execution(
+                    state,
+                    manager,
+                    input.clone(),
+                    observers,
+                    &exit_kind,
+                    false,
+                )?;
+
+                if feedback.is_interesting(state, manager, &input, observers, &exit_kind)? {
+                    // we found a new solution! use the smaller base
+                    base = input;
+
+                    // do more runs! maybe we can minify further
+                    next_i = 0;
+                }
+
+                corpus_idx
+            } else {
+                // we can't guarantee that the mutators provided will necessarily reduce size, so
+                // skip any mutations that actually increase size so we don't waste eval time
+                None
+            };
+
+            start_timer!(state);
+            self.mutator_mut().post_exec(state, i as i32, corpus_idx)?;
+            mark_feature_time!(state, PerfFeature::MutatePostExec);
+
+            i = next_i;
+        }
+
+        let mut hasher = AHasher::new_with_keys(0, 0);
+        base.hash(&mut hasher);
+        let new_hash = hasher.finish();
+        if base_hash != new_hash {
+            let mut testcase = Testcase::with_executions(base, *state.executions());
+            fuzzer
+                .feedback_mut()
+                .append_metadata(state, &mut testcase)?;
+            let prev = state.corpus_mut().replace(base_corpus_idx, testcase)?;
+            fuzzer
+                .scheduler_mut()
+                .on_replace(state, base_corpus_idx, &prev)?;
+        }
+
+        state.set_max_size(orig_max_size);
+
+        Ok(())
+    }
+}
+
+/// The default solution minimising mutational stage
+#[derive(Clone, Debug)]
+pub struct StdTMinMutationalStage<CS, E, EM, F1, F2, FF, I, M, S, T, Z>
+where
+    I: Input + HasLen,
+    M: Mutator<I, S>,
+{
+    mutator: M,
+    factory: FF,
+    runs: usize,
+    #[allow(clippy::type_complexity)]
+    phantom: PhantomData<(CS, E, EM, F1, F2, I, S, T, Z)>,
+}
+
+impl<CS, E, EM, F1, F2, FF, I, M, OT, S, Z> Stage<E, EM, S, Z>
```

Comment:
This is probably the most painful of the generics in this file. I don't want to believe that it has to be this complex :sweat_smile: Any ideas for reducing complexity here?

Reply:
I'm working on associated types in a branch

Reply:
But who knows if I manage to finish it, check out 
https://github.com/AFLplusplus/LibAFL/compare/main...associated_types

Reply:
That might be majorly API breaking :grimacing: But well worth the investment. Juggling 10+ types in a generic impl is mind-boggling, especially when you get just the wrong constraint (took me ~10 minutes to figure out that I had erroneously put `Z: HasObservers` while I was writing this).

Reply:
Yes that would be 0.9 *if* it works. Would help a lot though